Shell/Shuffling.hpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file Shuffling.hpp
+ * Defines class Shuffling implementing the shuffling of the input.
+ * @since 9/6/2021 Prague
+ */
+
+#ifndef __Shuffling__
+#define __Shuffling__
+
+#include "Forwards.hpp"
+
+#include "Lib/List.hpp"
+#include "Lib/Coproduct.hpp"
+#include "Kernel/Unit.hpp"
+
+#include <algorithm>
+
+namespace Shell {
+
+using namespace Lib;
+using namespace Kernel;
+
+/**
+ * Class implementing shuffling-related procedures.
+ * @since 09/06/2021 Prague
+ */
+class Shuffling
+{
+private:
+  typedef Coproduct<Formula*, Literal*, TermList> Shufflable;
+
+  static void shuffleIter(Shufflable sh);
+
+public:
+  static void polarityFlip(Problem&);
+
+  static void shuffle(Problem&);
+  static void shuffle(UnitList*&);
+
+  // shuffling should not require allocating new memory, so all the objects are modified "in place"
+
+  static void shuffle(Unit*);     // shuffle a unit; just dispatches to either shuffle(Clause*) or shuffle(Formula*)
+  static void shuffle(Clause*);   // shuffling a clause; it assumes literals are shared (i.e. nothing "special" in them anyway) so it does not touch those
+
+  static void shuffle(Formula* f) { shuffleIter(Shufflable(f)); }  // shuffling a formula; will try to descend also to the term level, to shuffle around commutative operators and formulas inside terms (if applicable)
+  static void shuffle(Literal* l) { shuffleIter(Shufflable(l)); }
+  static void shuffle(TermList tl) { shuffleIter(Shufflable(tl)); }
+
+  template<typename Arrayish>
+  static void shuffleArray(Arrayish& a, unsigned len) {
```

Comment:
This is Fisher-Yates, or something like it, right? Might be worth writing down exactly what it is for later Googling.

Reply:
Didn't know it was called that way. (Although I've heard of Knuth shuffle and of various ways of making mistakes in trying to draw a permutation uniformly at random.)